src/tool_getparam.c

```diff
@@ -2964,25 +2975,71 @@ ParameterError parse_args(struct GlobalConfig *global, int argc,
-        const char *nextarg = NULL;
-        if(i < (argc - 1)) {
-          nextarg = convert_tchar_to_UTF8(argv[i + 1]);
-          if(!nextarg) {
-            unicodefree(orig_opt);
-            return PARAM_NO_MEM;
+        const char *nextarg_param = NULL; /* Argument for getparameter */
+        const struct LongShort *longopt_match = NULL;
+        bool is_long_opt = ('-' == orig_opt[1]);
+
+        if(is_long_opt) {
+          const char *word = orig_opt + 2;
+          if(!strncmp(word, "no-", 3))
+            word += 3;
+          else if(!strncmp(word, "expand-", 7))
+            word += 7;
+          longopt_match = findlongopt(word);
+        }
+        else if(orig_opt[1] && !orig_opt[2]) { /* Single letter short option */
+          longopt_match = findshortopt(orig_opt[1]);
+        }
+        /* For combined short options like -sS, nextarg_param remains NULL
+           unless the last option in the sequence takes an argument.
+           getparameter handles parsing these combined options. */
+
+        /* Check if the option expects an argument and if nextarg is a valid candidate */
+        if(longopt_match && (ARGTYPE(longopt_match->desc) >= ARG_STRG)) {
+          if(i < (argc - 1)) {
+            const char *potential_arg = argv[i + 1];
+            /* If it's C_HTTP3, the argument is optional.
+               It's an argument if it doesn't start with '-' or is not NULL.
+               For other options requiring arguments, it's an argument if not NULL.
+            */
+            if(longopt_match->cmd == C_HTTP3) {
+              if(potential_arg && potential_arg[0] != '-') {
+                nextarg_param = convert_tchar_to_UTF8(potential_arg);
+                if(!nextarg_param) {
+                  unicodefree(orig_opt);
+                  return PARAM_NO_MEM;
+                }
+                passarg = TRUE;
+              }
+              /* If potential_arg is NULL or starts with '-', nextarg_param remains NULL,
+                 and passarg remains FALSE, indicating --http3 is used without a value or
+                 the next token is another option. */
+            }
+            else if (potential_arg) { /* Mandatory argument expected by other options */
+              nextarg_param = convert_tchar_to_UTF8(potential_arg);
+              if(!nextarg_param) {
+                  unicodefree(orig_opt);
+                return PARAM_NO_MEM;
+              }
+              passarg = TRUE;
+            }
           }
+          /* If i == argc - 1, nextarg_param remains NULL.
+             getparameter will handle PARAM_REQUIRES_PARAMETER for mandatory args.
+             For C_HTTP3, NULL nextarg_param is fine. */
         }
 
-        result = getparameter(orig_opt, nextarg, &passarg, global, config);
+        result = getparameter(orig_opt, nextarg_param, &passarg, global, config);
+        if(nextarg_param) /* if allocated by convert_tchar_to_UTF8 */
+          unicodefree(CURL_UNCONST(nextarg_param));
```

Comment:
isn't `unicodefree()` already doing `CURL_UNCONST()`, via `curlx_unicodefree()`? If so, this cast is redundant.